### DISCUSSION TITLE
Backup of SQLite includes content of -wal?

### DISCUSSION THREAD

**Discussion by ontheair81** [dani-garcia/vaultwarden#2369] [Q&A] [answered]
Hello, I am adding/editing entries in Vaultwarden every day, so `db.sqlite3-wal` is growing every day, while `db.sqlite3` stays the same size (as the database is not "[checkpointed](https://sqlite.org/wal.html)" [see chapter 2.1]).

Every night I am running a backup of the database as mentioned [here](https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#sqlite-database-files):
`sqlite3 data/db.sqlite3 ".backup '/path/to/backups/db-$(date '+%Y%m%d-%H%M').sqlite3'"`

I realized, that the backupped database file also has the same size every night. This seems odd, as there have been a lot of changes in the database.

So my questions are:
When doing a backup, is the content of  `db.sqlite3-wal` included in the backup? (I guess not.)
Does this mean, that a backup is just complete after the database has been checkpointed (or Vaultwarden has been stopped and restarted, as then checkpointing occurs)?
Does it make sense to do manual checkpoint prior to a backup (`PRAGMA wal_checkpoint`)?

Thanks for any help!

**Comment by jjlin** [comment, marked as answer]
The `.backup` command uses SQLite's [Online Backup API](https://sqlite.org/backup.html), which allows backup of a database without taking it "offline" (i.e., shutting down applications using the database). There's no need to manually checkpoint. A backup doesn't literally copy all of the contents of the WAL file, but as the backup operation copies the database page by page, the WAL file is [taken into account](https://sqlite.org/wal.html#concurrency) -- it would be a pretty useless backup otherwise.

**Comment by ontheair81** [comment]
Thank you very much for pointing out that the backup API takes the WAL into account, without doing a manual checkpoint!

In the meantime I created a test database on my desktop computer, experimenting with WAL and the backup API. This shows the same as what you said: Yes, the backup is complete, without checkpointing.

I was just worried about the fact, that my Vaultwarden backup files has been exactly the same size for a few days, although there has been changes in the database. As the content of this database is encrypted, I was not able to check if the new items are included, if course.
Still not sure why it was like this, but the last two backups now differ in size, which looks fine as there has been a lot of changes made in the content of Vaultwarden.

Thank you for your help and information, I will mark your post as answer!

**Comment by BlackDex** [comment]
Regarding the size not being changed.
Sqlite doesn't shrink the database if you delete items.
It just makes those records as deleted and it will overwrite those with new records.

If you want it to shrink you need to VACUUM the database which will kinda recreate the whole database and disregard the ones marked a deleted.

**Reply by ontheair81** [reply]
Good to know, this seems to be reason! Thank you for the explanation!
